REQUIREMENTS.md

```diff
@@ -0,0 +1,160 @@
+
+## Immutable
+
+IPLD links must be immutable. We'll likely define a mutable link spec on top of
+IPLD but there needs to be an immutable layer at the bottom.
+
+**Motivation:** *Having* an immutable layer is important for a lot of analysis,
+memoization, type checking, etc.
+
+## Multicodecs Are Not Types
+
+It's impossible to understand IPLD data at a *structural* level if we don't know
+the format. Therefore, we should avoid introducing new formats unnecessarily as
+*every* IPLD implementation needs to support these new formats.
+
+## No Non-Local Reasoning
+
+An IPLD block should never be interpreted in the context of *anything* not
+contained in the block (and CID).
+
+For example, assuming we add support for relative links, the following
+definition of `foo` would not be a valid IPLD block:
+
+```
+var foo = {
+  // points outside of the current block, into the parent's "baz" field.
+  "baz": {"/": "../../baz"}
+}
+var bar = {
+  "foo": CidOf(foo),
+  // `/foo/baz` points here.
+  "baz": "something"
+}
+
+// resolution throug block `foo` depends on block `bar`.
+Resolve("/ipld/${CidOf(bar)}/foo/baz/")
+```
+
+For the same reason, IPLD links can't rely on an authority (e.g., a blockchain).
+
+Note: Links like this can still be encoded at the application level but they
+won't be handled by the IPLD resolver (and won't get the special "link" type).
+
+**Motivation:** IPLD needs to be easy to reason about.
+
+**Caveat:**
+
+We *may* want to relax this if we want to move schemas into separate,
+deduplicated blocks (referenced by CID). If we do that, we'd need to fetch a
+block's schema before being able to interpret the it.
+
+However, we need to *thoroughly* discuss any changes to this requirement.
+
+1. The space savings may not be worth it given the size of CIDs (>40 bytes),
+   compression, smart transports, and smart datastores.
+2. This change would introduce some weird interface complexities and potential
+   network dependencies.
+
+## No Cycles
+
+IPLD links must not be cyclic, even if we add support for relative links.
+
+**Motivation:**
+
+1. Security: Forbidding link cycles ensures that any graph traversal terminates.
+   This makes it easier to correctly and securely implement some graph
+   algorithms, even on potentially untrusted data in a distributed system.
+2. Consistency: Without complex hacks, it's impossible to create link cycles
+   *between* hash-linked blocks. Allowing link cycles one one level (e.g., in
+   relative links within a single block) but not on a larger scale (between
+   blocks) is inconsistent (even though there's nothing we can do about it).
+
+## Stable Pathing
+
+An IPLD path always means the same thing, everywhere, every time. Importantly,
+users shouldn't be able to configure their IPLD library to change how path
+resolution works. They should be able to use alternative path resolution
+algorithms for *non*-IPLD paths (e.g., IPFS paths) but those should be built
+on top of IPLD.
```

Comment:
This is an interesting one.

I think some of our recent discussions include the idea that we might want to do "pathing" over data structures which use Advanced Data Layouts, such as HAMTs.  I also want to be able to "path" using struct field names when handling data with a Schema, even if the representation is tuple (e.g. becomes a list when encoded, and thus has indexes instead of string keys).  Nobody wants to dispute that "stable pathing" is important and useful, I think; but these other features are *also* distinctly important and useful too.  What to do?

Perhaps what we're seeing is another instance where we could benefit from a split, wherein the strictest form of this rule applies at the Data Model layer; and the spirit of the rule sticks around, but the technicalities are a bit relaxed with some of the other higher-level tools in the core ecosystem.

---

Another way to look at this is that the "alternative path resolution algorithms [...] (e.g., IPFS paths)" are now something that we can describe *slightly* more generally: e.g. we expect to be able to make "IPFS" paths work with a combination of the {IPLD Data Model + ADLs such as HAMT + a Schema saying where to expect it all}, and while that's then an "IPFS" application choice-set, it requires very very little glue code (in fact none: all this is 'declarative', except for any logic inside the ADL, and even that we're making spec-first/declarative-ish) to connect, since we've made these components reusable and composable through interfaces we had for IPLD anyway (e.g. Node).